src/bsp/src/stm32/src/SpiEsp.cpp

```diff
@@ -0,0 +1,211 @@
+#include "SpiEsp.h"
+#include "hal/esp_spi.h"
+#include "hal/hal_gpio.h"
+#include <FreeRTOSConfig.h>
+#include <cstring>
+
+void task(void* instance) {
+    const int loopRate = 30;
+    while (true) {
+        static_cast<SpiEsp*>(instance)->execute();
+        vTaskDelay(loopRate);
+    }
+}
+
+SpiEsp::SpiEsp(ICRC& crc, ILogger& logger) : m_crc(crc), m_logger(logger) {
+    m_txState = transmitState::IDLE;
+    m_rxState = receiveState::IDLE;
+    m_inboundMessage = {};
+    m_outboundMessage = {};
+    m_inboundRequest = false;
+    m_isBusy = false;
+
+    m_semaphore = xSemaphoreCreateBinaryStatic(&m_semaphoreBuffer);
+    xSemaphoreGive(m_semaphore);
+
+    setEspCallback(SpiEsp::espInterruptCallback, this);
+
+    m_taskHandle = xTaskCreateStatic(task, "esp_spi_driver_task", 1024u, this, 1u,
+                                     (StackType_t*)m_stackData.data(), &m_stackBuffer);
+}
+
+bool SpiEsp::send(const uint8_t* buffer, uint16_t length) {
+    bool retVal = false;
+
+    if (isBusy()) { // Not available
+    } else if (length >= ESP_SPI_MAX_MESSAGE_LENGTH) { // Message too long
+        m_logger.log(LogLevel::Warn,
+                     "SpiEsp: Message length of %d is larger than maximum allowed of %d", length,
+                     ESP_SPI_MAX_MESSAGE_LENGTH);
+    } else {
+
+        m_logger.log(LogLevel::Debug, "Sending message of length %d", length);
+        // TODO: this memcpy could change once we have a buffer manager/allocator of some sorts.
+        std::memcpy(m_outboundMessage.m_data.data(), buffer, length);
+        // Padding with 0 up to a word-alligned boundary
+        for (uint8_t i = 0; i < (length % 4); i++) {
+            m_outboundMessage.m_data[length] = 0;
+            length++;
+        }
+        // Appending CRC32
+        *(uint32_t*)(m_outboundMessage.m_data.data() + length) =
+            m_crc.calculateCRC32(m_outboundMessage.m_data.data(), length);
+        length += CRC32_SIZE;
+        m_outboundMessage.m_sizeBytes = length;
+        m_outboundMessage.m_sent = false;
+        m_txState = transmitState::SENDING_HEADER;
+        m_isBusy = true;
+        retVal = true;
+    }
+
+    return retVal;
+}
+
+bool SpiEsp::isBusy() const { return m_isBusy; }
+
+void SpiEsp::execute() {
```

Comment:
Maybe, but I am not sure what the gain would be. Both the processReception and sprocessSend would each need to be called in that order before a send_receive call. It would make for shorter functions but since the execute functions is already split into three distinct sections (rx switch case, tx switch case and send_receive call) I don't think it will really improve readability that much.